src/foundry/common/abstract/type-data.d.mts

```diff
@@ -81,21 +161,118 @@ declare class _InternalTypeDataModel<
 export default abstract class TypeDataModel<
   Schema extends DataSchema,
   Parent extends Document<DataSchema, any, any>,
-  BaseData extends Record<string, any> = Record<string, never>,
-  DerivedData extends Record<string, any> = Record<string, never>,
+  BaseData extends Record<string, unknown> = Record<string, never>,
+  DerivedData extends Record<string, unknown> = Record<string, never>,
 > extends _InternalTypeDataModel<Schema, Parent, BaseData, DerivedData> {
+  [__BaseModel]: DataModel<Schema, Parent>;
+  [__BaseData]: RemoveIndexSignatures<BaseData>;
+  [__DerivedData]: RemoveIndexSignatures<DerivedData>;
+
   modelProvider: System | Module | null;
 
   /**
-   * Prepare data related to this DataModel itself, before any derived data is computed.
+   * A set of localization prefix paths which are used by this data model.
    */
-  prepareBaseData(this: Merge<DataModel<Schema, Parent>, Partial<DerivedData>>): void;
+  static LOCALIZATION_PREFIXES: string[];
 
-  /* -------------------------------------------- */
+  /**
+   * Prepare data related to this DataModel itself, before any derived data is computed.
+   *
+   * Called before {@link ClientDocument#prepareBaseData} in {@link ClientDocument#prepareData}.
+   * */
+  prepareBaseData(this: TypeDataModel.PrepareBaseDataThis<this>): void;
 
   /**
    * Apply transformations of derivations to the values of the source data object.
    * Compute data fields whose values are not stored to the database.
+   *
+   * Called before {@link ClientDocument#prepareDerivedData} in {@link ClientDocument#prepareData}.
+   */
+  prepareDerivedData(this: TypeDataModel.PrepareDerivedDataThis<this>): void;
+
+  /**
+   * Convert this Document to some HTML display for embedding purposes.
+   * @param config  - Configuration for embedding behavior.
+   * @param options - The original enrichment options for cases where the Document embed content
+   *                  also contains text that must be enriched.
+   */
+  toEmbed(
+    config: TextEditor.DocumentHTMLEmbedConfig,
+    options: TextEditor.EnrichmentOptions,
+  ): Promise<HTMLElement | HTMLCollection | null>;
+
+  /* -------------------------------------------- */
+  /*  Database Operations                         */
+  /* -------------------------------------------- */
+
+  /**
+   * Called by {@link ClientDocument#_preCreate}.
+   *
+   * @param data    - The initial data object provided to the document creation request
+   * @param options - Additional options which modify the creation request
+   * @param user    - The User requesting the document creation
+   * @returns Return false to exclude this Document from the creation operation
+   */
+  protected _preCreate(
+    data: fields.SchemaField.AssignmentType<Schema>,
```

Comment:
These types for _preCreate through _onUpdate aren't quite right - it's the entire *parent's* assignment type, but we have a unique chance to write a union here for something like
```ts
data: fields.SchemaField.AssignmentType<ParentSchema> & { system: fields.SchemaField.AssignmentType<Schema> }
```

Can we use `infer` in the top level generic to get ParentSchema?